chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -830,3 +870,137 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
+		}
+		chainntnfs.Log.Infof("Catching up on missed block w/ height %d", height)
+		if err := b.handleBlockConnected(height, *hash); err != nil {
+			return fmt.Errorf("error on handling connected block: %v",
+				err)
+		}
+	}
+	return nil
+}
+
+// rewindChain handles internal state updates associated with disconnected blocks.
+// It has no effect if given a height greater than or equal to our current best
+// known height.
+func (b *BitcoindNotifier) rewindChain(newBestHeight int32) error {
+	for height := b.bestBlock.Height; height > newBestHeight; height-- {
+		hash, err := b.chainConn.GetBlockHash(int64(height - 1))
+		if err != nil {
+			return fmt.Errorf("unable to find blockhash for disconnected height=%d: %v",
+				height, err)
+		}
+
+		chainntnfs.Log.Infof("Block disconnected from main chain: "+
+			"height=%v, sha=%v", height, hash)
+
+		err = b.txConfNotifier.DisconnectTip(uint32(height))
+		if err != nil {
+			return fmt.Errorf("unable to disconnect tip for height=%d: %v",
+				height, err)
+		}
+		// Since we've successfully disconnected, set new best block.
+		b.bestBlock.Height = height - 1
+		b.bestBlock.Hash = hash
+	}
+	return nil
+}
+
+// catchUpClientOnBlocks is called when a notification client's best block
+// is behind our current best block.
+// It dispatches block notifications for all the blocks missed by the client
+// or does nothing if the client doesn't send us its best block.
+func (b *BitcoindNotifier) catchUpClientOnBlocks(bestBlock *chainntnfs.BlockEpoch) error {
+	if bestBlock == nil {
+		return nil
+	}
+	if bestBlock.Hash == nil {
+		return fmt.Errorf("can't catch up client on blocks: best block " +
+			"hash is nil")
+	}
+
+	// If a reorg causes the client's best hash to be incorrect, retrieve the
+	// closest common ancestor and dispatch notifications from there.
+	hashAtBestHeight, err := b.chainConn.GetBlockHash(int64(bestBlock.Height))
+	if err != nil {
+		return fmt.Errorf("unable to find blockhash for height=%d: %v",
+			bestBlock.Height, err)
+	}
+	commonAncestorHeight, err := b.getCommonBlockAncestorHeight(
+		*bestBlock.Hash,
+		*hashAtBestHeight,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to find common ancestor: %v", err)
+	}
+
+	numMissedBlocks := b.bestBlock.Height - commonAncestorHeight
+	if numMissedBlocks < 0 {
+		numMissedBlocks = 0
+	}
+	for height := commonAncestorHeight + 1; height <= b.bestBlock.Height; height++ {
+		hash, err := b.chainConn.GetBlockHash(int64(height))
+		if err != nil {
+			return fmt.Errorf("unable to find blockhash for height=%d: %v",
+				height, err)
+		}
+		b.notifyBlockEpochs(height, hash)
+	}
+	return nil
+}
+
+// getCommonBlockAncestorHeight takes in:
+// (1) the hash of a block that has been reorged out of the main chain
+// (2) the hash of the block of the same height from the main chain
+// It returns the height of the nearest common ancestor between the two hashes,
+// or an error
+func (b *BitcoindNotifier) getCommonBlockAncestorHeight(
```

Comment:
Since this function doesn't modify any internal state, this could be moved into the `chainntnfs` package and exposed as a helper method. I think btcd and bitcoind can share this method since their `chainConn`'s both implement `chain.Interface` as:
```golang
height, err := chainntnfs.GetCommonBlockAncestorHeight(
        b.chainConn, reorgHash, chainHash,
)
```

Reply:
Lol, love dis idea. I think `btcd`'s `chainConn` might not, though. Looks like `rpcclient.Client` is missing a few of the `chain.Interface`'s methods?